service/topology_coordinator.cc

```diff
@@ -1551,6 +1554,85 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                    // and then a coordinator again, it's fine - we'll just repeat this step.
+                    // (If we're in `left` state when we try to restart we won't
+                    // be able to become a voter - we'll be banned from the cluster.)
+                }
+
+                bool barrier_failed = false;
+                // Wait until other nodes observe the new token ring and stop sending writes to this node.
+                try {
+                    node = retake_node(co_await global_token_metadata_barrier(std::move(node.guard), get_excluded_nodes(node)), node.id);
+                } catch (term_changed_error&) {
+                    throw;
+                } catch (group0_concurrent_modification&) {
+                    throw;
+                } catch (...) {
+                    rtlogger.error("transition_state::left_token_ring, "
+                                    "raft_topology_cmd::command::barrier failed, error {}",
+                                    std::current_exception());
+                    barrier_failed = true;
+                }
+
+                if (barrier_failed) {
+                    // If barrier above failed it means there may be unfinished writes to a decommissioned node.
+                    // Lets wait for the ring delay for those writes to complete and new topology to propagate
+                    // before continuing.
+                    co_await sleep_abortable(_ring_delay, _as);
+                    node = retake_node(co_await start_operation(), node.id);
+                }
+
+                topology_request_tracking_mutation_builder rtbuilder(node.rs->request_id);
+
+                rtbuilder.done();
+
+                co_await update_topology_state(take_guard(std::move(node)), {rtbuilder.build()}, "report request completion in left_token_ring sate");
+
+                // Tell the node to shut down.
+                // This is done to improve user experience when there are no failures.
+                // In the next state (`node_state::left`), the node will be banned by the rest of the cluster,
+                // so there's no guarantee that it would learn about entering that state even if it was still
+                // a member of group0, hence we use a separate direct RPC in this state to shut it down.
+                //
+                // There is the possibility that the node will never get the message
+                // and decommission will hang on that node.
+                // This is fine for the rest of the cluster - we will still remove, ban the node and continue.
+                auto node_id = node.id;
+                bool shutdown_failed = false;
+                try {
+                    node = co_await exec_direct_command(std::move(node), raft_topology_cmd::command::barrier);
+                } catch (...) {
+                    rtlogger.warn("failed to tell node {} to shut down - it may hang."
```

Comment:
So how does it work after d576ed31dce292997d1cf32af5a9e89768b154d7? The node no longer waits for shutdown RPC, and shuts down automatically once it observes request completion? In which case there is a race here (which shutdown RPC was precisely designed to avoid)  -- the node may shutdown before we round-trip with the barrier here in which case we'll report a confusing warning. @gleb-cloudius

Reply:
And the comment is outdated.

Reply:
Gleb I vaguely recall that you attempted to remove the shutdown RPC once, and I think it was @piodul who catched it during review pointing out the race. Looks like you finally got your way :(

Reply:
> Gleb I vaguely recall that you attempted to remove the shutdown RPC once, and I think it was @piodul who catched it during review pointing out the race. Looks like you finally got your way :(

The original purpose of the shutdown RPC was to not fail requests due to decomissioning node shutting down before drain completes in `left_token_ring` state. My previous attempt to remove it returned the race and that is what caught during the review. Now this may not happen any longer since a decommissioning node no longer  relies on the node state to decide when operation completes and we mark the operation as completed only after all nodes observed the global metadata barrier.

The node may exit before this barrier here, but the warning is not confusing. In fact it is correct. It says the node may hand, not that it hanged.